lib/diff.js

```diff
@@ -67,86 +64,96 @@ class Diff extends BaseCommand {
       )
     }
 
+    // diffWorkspaces may have set this already
+    if (!this.prefix)
+      this.prefix = this.npm.prefix
+
+    // this is the "top" directory, one up from node_modules
+    // in global mode we have to walk one up from globalDir because our
+    // node_modules is sometimes under ./lib, and in global mode we're only ever
+    // walking through node_modules (because we will have been given a package
+    // name already)
+    if (this.npm.config.get('global'))
+      this.top = resolve(this.npm.globalDir, '..')
+    else
+      this.top = this.prefix
+
     const [a, b] = await this.retrieveSpecs(specs)
     npmlog.info('diff', { src: a, dst: b })
 
-    const res = await libdiff([a, b], {
+    const res = await libnpmdiff([a, b], {
       ...this.npm.flatOptions,
       diffFiles: args,
-      where: this.where,
+      where: this.top,
     })
     return this.npm.output(res)
   }
 
-  async retrieveSpecs ([a, b]) {
-    // no arguments, defaults to comparing cwd
-    // to its latest published registry version
-    if (!a)
-      return this.defaultSpec()
-
-    // single argument, used to compare wanted versions of an
-    // installed dependency or to compare the cwd to a published version
-    if (!b)
-      return this.transformSingleSpec(a)
-
-    const specs = await this.convertVersionsToSpecs([a, b])
-    return this.findVersionsByPackageName(specs)
+  async diffWorkspaces (args, filters) {
+    const workspaces =
+      await getWorkspaces(filters, { path: this.npm.localPrefix })
+    for (const workspacePath of workspaces.values()) {
+      this.top = workspacePath
+      this.prefix = workspacePath
+      await this.diff(args)
+    }
   }
 
-  async defaultSpec () {
-    let noPackageJson
-    let pkgName
+  // get the package name from the packument at `path`
+  // throws if no packument is present OR if it does not have `name` attribute
+  async packageName (path) {
+    let name
     try {
-      pkgName = await readPackageName(this.npm.prefix)
+      // TODO this won't work as expected in global mode
```

Comment:
I believe this use case is already handled in L114 throwing an EUSAGE error

Reply:
No, read-package-json no longer throws in global mode. That check didn't belong in there.

Reply:
oohh I see what you mean 👍